core/src/main/scala/org/apache/spark/sql/delta/files/DeltaSourceSnapshot.scala

```diff
@@ -52,51 +54,45 @@ class DeltaSourceSnapshot(
     (part, data)
   }
 
-  protected def initialFiles: Dataset[IndexedFile] = {
+  lazy val initialFiles: DataFrame = {
     import spark.implicits.rddToDatasetHolder
     import org.apache.spark.sql.delta.implicits._
 
+    val numPartitions = snapshot.getNumPartitions
+
     cacheDS(
-      snapshot.allFiles.sort("modificationTime", "path")
+      snapshot.allFiles
+        // This allows us to control the number of partitions created from the sort instead of
+        // using the shufflePartitions setting
+        .repartitionByRange(numPartitions, col("modificationTime"), col("path"))
```

Comment:
Given the perf questions, it might be better to save this particular change for a separate PR.
Unless the range partitioning is required for this PR to achieve the desired effect?
(but I'd at least _hope_ spark's normal settings would not lead to OOM)?

Reply:
The sort is already doing a repartitionByRange, this is simply a mechanism to control the number of partitions output from that. Otherwise the sort will do a repartition by range to spark.sql.shuffle.partitioms

Reply:
```
>>> spark.createDataFrame([['a']], ['a']).sort('a').explain()
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Sort [a#30 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(a#30 ASC NULLS FIRST, 200), ENSURE_REQUIREMENTS, [plan_id=66]
      +- Scan ExistingRDD[a#30]


>>> spark.createDataFrame([['a']], ['a']).repartitionByRange(10, 'a').sort('a').explain()
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Sort [a#32 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(a#32 ASC NULLS FIRST, 10), REPARTITION_BY_NUM, [plan_id=77]
      +- Scan ExistingRDD[a#32]
```

Reply:
The reason it's relevant/related to this PR is that if you let the spark.sql.shuffle.partitions take effect, you can end up with tens of thousands of partitions in the final state we are `toLocalterator`ing, which will then have to trigger tens of thousands of jobs just to get through the snapshot.

Reply:
We routinely have it set in the tens of thousands for our large jobs